src/ethereum/crypto/finite_field.py

```diff
@@ -0,0 +1,399 @@
+        ...
+
+    def __imul__(self: F, right: F) -> F:
+        ...
+
+    def __pow__(self: F, exponent: int) -> F:
+        ...
+
+    def __ipow__(self: F, right: int) -> F:
+        ...
+
+    def __neg__(self: F) -> F:
+        ...
+
+    def __truediv__(self: F, right: F) -> F:
+        ...
+
+
+T = TypeVar("T", bound="PrimeField")
+
+
+class PrimeField(int, Field):
+    """
+    Superclass for integers modulo a prime. Not intended to be used
+    directly, but rather to be subclassed.
+    """
+
+    __slots__ = ()
+    PRIME: int
+
+    @classmethod
+    def from_be_bytes(cls: Type, buffer: "Bytes") -> "Uint":
+        """
+        Converts a sequence of bytes into a element of the field.
+        Parameters
+        ----------
+        buffer :
+            Bytes to decode.
+        Returns
+        -------
+        self : `Uint`
+            Unsigned integer decoded from `buffer`.
+        """
+        return cls(int.from_bytes(buffer, "big"))
+
+    @classmethod
+    def zero(cls: Type[T]) -> T:
+        return cls.__new__(cls, 0)
+
+    @classmethod
+    def from_int(cls: Type[T], n: int) -> T:
+        return cls(n)
+
+    def __new__(cls: Type[T], value: int) -> T:
+        return int.__new__(cls, value % cls.PRIME)
+
+    def __radd__(self: T, left: T) -> T:  # type: ignore[override]
+        return self.__add__(left)
+
+    def __add__(self: T, right: T) -> T:  # type: ignore[override]
+        if not isinstance(right, int):
+            return NotImplemented
+
+        return self.__new__(type(self), int.__add__(self, right))
+
+    def __iadd__(self: T, right: T) -> T:  # type: ignore[override]
+        return self.__add__(right)
+
+    def __sub__(self: T, right: T) -> T:  # type: ignore[override]
+        if not isinstance(right, int):
+            return NotImplemented
+
+        return self.__new__(type(self), int.__sub__(self, right))
+
+    def __rsub__(self: T, left: T) -> T:  # type: ignore[override]
+        if not isinstance(left, int):
+            return NotImplemented
+
+        return self.__new__(type(self), int.__rsub__(self, left))
+
+    def __mul__(self: T, right: T) -> T:  # type: ignore[override]
+        if not isinstance(right, int):
+            return NotImplemented
+
+        return self.__new__(type(self), int.__mul__(self, right))
+
+    def __rmul__(self: T, left: T) -> T:  # type: ignore[override]
+        return self.__mul__(left)
+
+    def __imul__(self: T, right: T) -> T:  # type: ignore[override]
+        return self.__mul__(right)
+
+    __floordiv__ = None  # type: ignore
+    __rfloordiv__ = None  # type: ignore
+    __ifloordiv__ = None
+    __divmod__ = None  # type: ignore
+    __rdivmod__ = None  # type: ignore
+
+    def __pow__(self: T, exponent: int) -> T:  # type: ignore[override]
+        # We should be able to omit the modulus here in Python3.8. This would
+        # use the Euclidian Algorithm and be considerably faster, but does not
+        # not work under PyPy.
```

Comment:
pypy v7.3.8 is out, and supports Python 3.9.

If it's significantly better, we could bump our minimum pypy version.

Reply:
We should probably bump our minimum pypy version. The current hack is perfectly acceptable in the meantime.

I feel the Python developers have had plenty of time to implement this given that the [basic idea](https://en.wikipedia.org/wiki/Euclidean_algorithm) was described by Euclid 2300 years ago and it has been applied to this problem since [at least the 6th century A.D](https://hsm.stackexchange.com/questions/7383/who-extended-the-euclidean-algorithm-to-derive-the-bezout-identity). My hack to work round this deficiency is comparatively recent, dating [from 1640](https://en.wikipedia.org/wiki/Fermat%27s_little_theorem).